zfs-replicate.sh

```diff
@@ -1,524 +1,538 @@
+        $ZFS snapshot -r ${local_set}@${sname}
+      fi
+    elif [ ${MODE} = PULL ] && [ ${TYPE} = LOCAL ]; then
+      printf "Creating ZFS snapshot %s@%s\n" "${remote_set}" "${sname}"
+      if [ $RECURSE_CHILDREN -ne 1 ]; then
+        printf "RUNNING: %s snapshot %s@%s\n" "${ZFS}" "${remote_set}" "${sname}"
+        ${ZFS} snapshot ${remote_set}@${sname}
+      else
+        printf "RUNNING: %s snapshot -r %s@%s\n" "${ZFS}" "${remote_set}" "${sname}"
+        ${ZFS} snapshot -r ${remote_set}@${sname}
+      fi
+    fi
+
+    ## check return
+    local status=$?
+    if [ $status -ne 0 ]; then
+      exit_clean $status
+    fi
+
+    ## Send incremental snaphot if count is 1 or more, otherwise send full snapshot
+    if [ ${MODE} == PUSH ]; then
+      if [ $scount -ge 1 ]; then
+        if ! do_push ${base_snap} ${local_set}@${sname} ${remote_set}; then
+          do_destroy ${local_set}@${sname}
+          exit_clean 99 ## TODO: use proper status
         fi
-        ## Send incremental snaphot if count is 1 or more, otherwise send full snapshot
-        if [ ${MODE} == PUSH ]; then    
-            if [ $scount -ge 1 ]; then
-                if ! do_push ${base_snap} ${local_set}@${sname} ${remote_set}; then
-		do_destroy ${local_set}@${sname}
-		exit_error
-		fi
-            else
-                if ! do_push "NULL" ${local_set}@${sname} ${remote_set}; then
-		do_destroy ${local_set}@${sname}
-		exit_error
-		fi
-            fi
-        elif [ ${MODE} == PULL ]; then
-            if [ $scount -ge 1 ]; then
-		if ! do_pull ${base_snap} ${remote_set}@${sname} ${local_set}; then
-		do_destroy ${remote_set}@${sname}
-		exit_error
-		fi
-            else
-                if ! do_pull "NULL" ${remote_set}@${sname} ${local_set}; then
-		do_destroy ${remote_set}@${sname}
-		exit_error
-		fi
-            fi
+      else
+        if ! do_push "NULL" ${local_set}@${sname} ${remote_set}; then
+          do_destroy ${local_set}@${sname}
+          exit_clean 99 ## TODO: use proper status
         fi
-        ## check return of replication
-        if [ $? != 0 ]; then
-            if [ ${MODE} = PUSH ]; then
-                printf "ERROR: Failed to send snapshot %s@$%s\n" "${local_set}" "${sname}"
-                printf "Deleting the local snapshot %s@$%s\n" "${local_set}" "${sname}"
-                do_destroy ${local_set}@${sname}
-                exit_error
-            elif [ ${MODE} = PULL ]; then
-                printf "ERROR: Failed to send snapshot %s@$%s\n" "${remote_set}" "${sname}"
-                printf "Deleting the local snapshot %s@$%s\n" "${remote_set}" "${sname}"
-                do_destroy ${remote_set}@${sname}
-                exit_error
-            fi
+      fi
+    elif [ ${MODE} == PULL ]; then
+      if [ $scount -ge 1 ]; then
+        if ! do_pull ${base_snap} ${remote_set}@${sname} ${local_set}; then
+          do_destroy ${remote_set}@${sname}
+          exit_clean ## TODO: use proper status
```

Comment:
Same issure with `exit_clean` here and for every function. The `exit_clean` should only be used when the replication completes successfully.

Reply:
It looks like it merged the exit codes into one block, so maybe it will work ok, but I don still want a FAILED when it fails the replication and SUCCESS when it completes.